### PR TITLE
Redis based cache

### PR DESCRIPTION
**Pull request by asze17**

### CODE REVIEW

File: src/bincode_impl.rs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    {
+        let v = bincode::encode_to_vec(&self.0, CONFIG).expect("failed to serialize item");
+
+        out.write_arg(&v);
+    }
+}
+
+impl<T: Decode> FromRedisValue for BincodeType<T> {
+    fn from_redis_value(v: &Value) -> deadpool_redis::redis::RedisResult<Self> {
+        match v {
```
Comment (jay3332) [inline src/bincode_impl.rs]: can be replaced with
 ```rs
if let Value::Data(d) = v { ... } else { ... }
```

File: src/cache.rs [hunk 2]
```diff
@@ -1,133 +1,268 @@
...
+
+use crate::{
+    bincode_impl::BincodeType,
+    error::Result,
+    models::{ChannelType, Permissions, User, UserFlags},
+};
+
+static POOL: OnceLock<Pool> = OnceLock::new();
+
+type ResultOption<T> = Result<Option<T>>;
```
Comment (jay3332) [inline src/cache.rs]: what the act

File: src/db/guild.rs [hunk 3]
```diff
@@ -49,30 +49,19 @@ pub trait GuildDbExt<'t>: DbExt<'t> {
...
-            read_lock
-                .existing_guild_ids
-                .as_ref()
-                .map(|g| g.contains(&guild_id))
-        };
+        let guild_cached = cache::guild_exist(guild_id).await?;
 
-        let guild_exists = if let Some(contains) = guild_cached {
-            contains
+        let guild_exists = if guild_cached.is_some() {
```
Comment (jay3332) [inline src/db/guild.rs]: why not use `Option::or`

File: src/db/guild.rs [hunk 4]
```diff
@@ -121,24 +106,15 @@ pub trait GuildDbExt<'t>: DbExt<'t> {
...
-            read_lock
-                .guild(guild_id)
-                .and_then(|g| g.members.as_ref())
-                .map(|m| m.contains(&user_id))
-        };
+        let cached = cache::is_member_of_guild(guild_id, user_id).await?;
 
-        let member_in_guild = if let Some(contains) = cached {
-            contains
+        let member_in_guild = if cached.is_some() {
```
Comment (jay3332) [inline src/db/guild.rs]: why not option::or

File: src/models/guild.rs [hunk 5]
```diff
@@ -160,3 +161,4 @@ bitflags::bitflags! {
 }
 
 serde_for_bitflags!(u32: GuildFlags);
+bincode_for_bitflags!(GuildFlags);
```
Comment (jay3332) [inline src/models/guild.rs]: such can be combined with serde_for_bitflags!

**Review by jay3332 — CHANGES_REQUESTED** [review]
YOUR NAME, CRYPTE

File: src/bincode_impl.rs (hunk 1 above)
Comment (asze17) [reply src/bincode_impl.rs]: k

File: src/cache.rs (hunk 2 above)
Comment (asze17) [reply src/cache.rs]: jayse

File: src/db/guild.rs (hunk 3 above)
Comment (asze17) [reply src/db/guild.rs]: ok

File: src/db/guild.rs (hunk 4 above)
Comment (asze17) [reply src/db/guild.rs]: forgot

File: src/models/guild.rs (hunk 5 above)
Comment (asze17) [reply src/models/guild.rs]: feel free

File: src/lib.rs [hunk 6]
```diff
@@ -104,9 +130,18 @@ macro_rules! serde_for_bitflags {
         );
     };
 
-    (u32: $t:ty) => { serde_for_bitflags!(@serde_unsigned(u32) $t => Int32); };
-    (i16: $t:ty) => { serde_for_bitflags!(@serde_signed(i16) $t => Int32); };
-    (i64: $t:ty) => { serde_for_bitflags!(@serde_signed(i64) $t => Int64); };
+    (u32: $t:ty) => {
+        serde_for_bitflags!(@serde_unsigned(u32) $t => Int32);
+        serde_for_bitflags!(@bincode for $t);
```
Comment (jay3332) [inline src/lib.rs]: put this after line 115

**Review by jay3332 — CHANGES_REQUESTED** [review]
act

**Pull request merged**